src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -244,6 +262,43 @@ CHIP_ERROR CodegenDataModelProvider::ServerClusters(EndpointId endpointId,
     VerifyOrReturnValue(endpoint->clusterCount > 0, CHIP_NO_ERROR);
     VerifyOrReturnValue(endpoint->cluster != nullptr, CHIP_NO_ERROR);
 
+    // We build the cluster list by merging two lists:
+    //   - mRegistry items from ServerClusterInterfaces
+    //   - ember metadata clusters
+    //
+    // This is done because `ServerClusterInterface` allows full control for all its metadata,
+    // in particular `data version` and `flags`.
+    //
+    // To allow cluster implementations to be incrementally converted to storing their own data versions,
+    // instead of relying on the out-of-band emberAfDataVersionStorage, first check for clusters that are
+    // using the new data version storage and are registered via ServerClusterInterfaceRegistry, then fill
+    // in the data versions for the rest via the out-of-band mechanism.
+
+    // assume the clusters on endpoint does not change in between these two loops
+    auto clusters               = mRegistry.ClustersOnEndpoint(endpointId);
+    size_t registryClusterCount = 0;
+    for ([[maybe_unused]] auto _ : clusters)
```

Comment:
Is there a reason ClustersList doesn't have a size()?

I.e. even if it's a linked list now, it could be optimized in the future to track size, so centralizing this behavior there seems worthwhile

Reply:
It is expensive to get now, so I did not implement it directly to discourage its use too frequently. Unfortunately we directly need it now. I can also do `std::distance(clusters.begin(), clusters.end())` and that can auto-optimize if we have a correct iterator.

That being said however: this is codegen which we will switch to something else. So the reality is that this is likely never optimized further or changed ... this will always stay a linked list until we maybe delete it completely at some time. So I would like to YAGNI optimizations or members  out.